### PR TITLE
Support partial content in media handler

### PR DESCRIPTION
**Pull request by rkfg**
I've added Range header support, this fixes vector-im/riot-web#4265

I might not have covered all the edge cases, it works for me but if this solution is too crude or suboptimal (for example, it always sends the file up to the last byte), feel free to reject the PR.

### CODE REVIEW

**Comment by richvdh** [comment]
@matrixbot: test this please

File: synapse/rest/media/v1/_base.py [hunk 1]
```diff
@@ -76,7 +76,40 @@ def respond_with_file(request, media_type, file_path,
...
+            if rangeStr:
+                parts = rangeStr.split('=')
+                if len(parts) == 2 and parts[0].lower() == "bytes":
+                    range = parts[1].split('-')
+                    if len(range) == 2:
+                        try:
+                            rangeFrom = int(range[0])
+                            if rangeFrom >= file_size or rangeFrom < 0:
+                                raise Exception
+                            f.seek(rangeFrom)
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]: I think this is a thing that could block, so probably shouldn't happen in the foreground thread.

File: synapse/rest/media/v1/_base.py [hunk 2]
```diff
@@ -76,7 +76,40 @@ def respond_with_file(request, media_type, file_path,
 
         add_file_headers(request, media_type, file_size, upload_name)
 
+        request.setHeader(
+            b"Accept-Ranges", b"bytes"
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]: no need to break this onto multiple lines

File: synapse/rest/media/v1/_base.py [hunk 3]
```diff
@@ -76,7 +76,40 @@ def respond_with_file(request, media_type, file_path,
...
+                    range = parts[1].split('-')
+                    if len(range) == 2:
+                        try:
+                            rangeFrom = int(range[0])
+                            if rangeFrom >= file_size or rangeFrom < 0:
+                                raise Exception
+                            f.seek(rangeFrom)
+                            request.setHeader(
+                                b"Content-Length",
+                                b"%d" % (file_size - rangeFrom)
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]: our coding style asks for a trailing comma here

**Comment by rkfg** [comment]
Uhh, so basically each line is wrong in some aspect. I think it's better to rewrite this completely for someone else who's on more friendly terms with python. And actually it *does* fix the aforementioned bug because I can now jump into the middle of the video and this function *does* serve media content (at least videos). It would be weird to submit a pull request that doesn't solve the issue while I explicitly stated it does.

So I think this should be left as a PoC and closed until someone impements the same idea in a better code.

**Comment by richvdh** [comment]
> It would be weird to submit a pull request that doesn't solve the issue while I explicitly stated it does.

You'd be surprised how often people do exactly that sort of thing... 

I'm completely mystified as to how it works, though. I just can't see anywhere that function is called outside the thumbnail handler.

**Comment by rkfg** [comment]
Well, it was removed in the develop branch. I actually was patching against 0.26.0, seems that this function usage has been removed in dd3092c3a357be2453e25293b5590d627f3cfc48, right after 0.26.0 was released. Never mind then, I hope this will be fixed in MediaStorage.

**Pull request closed without merging**